File: cs/Problems/Day1.cs
namespace AOC2023.Problems;

public class Day1 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => SumFirstAndLastDigits(input);

    private static int SumFirstAndLastDigits(ReadOnlySpan<char> input)
    {
        Span<Range> gameLines = stackalloc Range[1000];
        int sum = 0;

        int lines = input.Split(gameLines, InputReader.NewLine);
        for (int c = 0; c < lines; c++)
        {
            var line = input[gameLines[c]];
            (int d1, int d2) = (0, 0);

            for (int i = 0, j = line.Length - 1; j >= 0 && (d1 == 0 || d2 == 0); i++, j--)
            {
                if (d1 == 0 && char.IsDigit(line[i]))
                    d1 = line[i] - '0';

                if (d2 == 0 && char.IsDigit(line[j]))
                    d2 = line[j] - '0';
            }

            sum += d1 * 10 + d2;
        }

        return sum;
    }
}

File: go/problems/day1.go
package problems

import (
	"strconv"
	"strings"
)

type Day1 struct{}

func (d Day1) Solve(input string) int {
	return sumFirstAndLastDigits(input)
}

func sumFirstAndLastDigits(input string) int {
	sum := 0
	for _, line := range strings.Split(input, NewLine) {
		d1, d2 := 0, 0

		for i, j := 0, len(line)-1; j >= 0 && (d1 == 0 || d2 == 0); i, j = i+1, j-1 {
			if d1 == 0 {
				if n, err := strconv.Atoi(line[i : i+1]); err == nil {
					d1 = n
				}
			}
			if d2 == 0 {
				if n, err := strconv.Atoi(line[j : j+1]); err == nil {
					d2 = n
				}
			}
		}

		sum += d1*10 + d2
	}
	return sum
}
